### DISCUSSION TITLE
linking uppercase and lowercase for a logographic script purpose

### DISCUSSION THREAD

**Discussion by lenivn** [fontforge/fontforge#5239] [Q&A]
Hi, i'm really newbie in fontforge and know nothing about programming. Quick summary, i'm creating a logographic script using ligatures as my main trick but the thing gets complicated when I use multiple glyphs for one ligature. To illustrate this in simple words, when I type the word "dragon" (= ligature "d_r_a_g_o_n") it's replaced by one logograph/glyph that contain all the letters but express only that word. Similar like Japanese typing inputs work. But when I use uppercasse glyphs things get messy and I have to create an almost infinite amount of ligatures "D_R_A_G_O_N", "D_r_a_g_o_n", "d_R_a_g_o_n", "D_r_A_G_o_N", etc, etc, etc... so my question is... Does exist an easy way of coding or promt to link upper and lowercase so i don't have to use all the possible ligatures? if you can explain me step by step would be wonderful, I haven't find any tutorial outside yet and English isn't my first language so it's even harder for me to find something useful. For the moment, I'm only using small caps in my script to avoid all nightmarish ligatures but I'd be glad if you all could help me.

**Comment by skef** [comment]
To be clear, if someone types dragon, DRAGON, or DrAGoN you want the same thing to appear in each case, or different things?

**Reply by lenivn** [reply]
Same thing for each case. If I type all the three ways (or infinite amount of variants between upper and lowercase) it has to appear one glyph. DRAGON / dragon / DrAgOn = 龍

**Reply by skef** [reply]
Hmm. I know how one would do this with character classes in a feature file but I'm not sure about the FontForge GUI. @ctrlcctrlv any advice (beyond switching to feature files)?

**Reply by lenivn** [reply]
I did a quick Google search trying to understand what u meant 😅 As I said, I'm not a programmer so I'm not familiarized with random terms. Is there any way to export a feature file to Fontforge?

**Reply by skef** [reply]
One can *import* a feature file *into* FontForge or *export* a feature file *out of* FontForge, but often with complicated fonts one switches from using the FontForge GUI for substitutions to using a feature file directly. 

The adobe feature file format is documented here: http://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html 

However, an explanation of its use is way beyond the scope of our discussion forums here FontForge. Fonts are complicated!

**Comment by skef** [comment]
Just learned something unfortunate for your needs:

https://github.com/adobe-type-tools/afdko/blob/develop/docs/OpenTypeFeatureFileSpecification.md#5d-gsub-lookuptype-4-ligature-substitution

OpenType doesn't support ligature substitution with character classes. That means that even if one can express what you want with a feature file using classes, the substitutions will be expanded out. Therefore substituting the equivalent of `sub [d D] [r R] [a A] [g G] [o O] [n N] by \dragon` will be expanded out into 64 rules. That's probably fine for just the one word, but if you plan to do this with many words you may effectively run out of space.

**Reply by lenivn** [reply]
It's fine. I already manage to resolve that, just don't know how. But I did it. Thanks for your help either way the link for the feature file format was really helpful :D
Right now, I'm just having new problems with some other glyphs related to "kanji" stuff.
